src/file-selector.ts

```diff
@@ -147,64 +147,63 @@ function fromDataTransferItem(
   return Promise.resolve(fwp);
 }
 
-// https://developer.mozilla.org/en-US/docs/Web/API/FileSystemEntry
-async function fromEntry(entry: any) {
-  return entry.isDirectory ? fromDirEntry(entry) : fromFileEntry(entry);
+async function fromEntry(
+  entry: FileSystemEntry,
+): Promise<FileWithPath | FileArray[]> {
+  if (isFileSystemDirectoryEntry(entry)) {
```

Comment:
I'm not sure I agree with this direction. We turned a simple 1 line function into 5+ lines and introduced 2 additional function calls (if the 1st condition is not met).

And it's not particularly more readable. It's in fact just more characters to read and process.

Reply:
Putting the [type predicate](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates) in a single function allows encapsulating possibly unsafe code, I could also type-cast it after the condition, but that would make it more opaque. I honestly don't think writing more code is problematic if it adds more guarantees about correctness.

Reply:
In JavaScript, there are no guarantees ... these types are still just things we tell the compiler we expect, which may not necessarily be true at runtime.

Again, just for the sake of enforcing types, I wouldn't change the logic. For correctness and soundness, I'd just make sure we have good coverage and integration tests (which we don't have at the moment).

Reply:
I'll resolve this thread, as we're essentially discussing this in the thread above already and this just fragments the discussion.